[Jason Ray] iP

JasonRay168 opened the pull request:
EchoBot
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

EchoBot frees your mind of having to remember things you need to do. It's,

- Text-based
- Easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. Download it
2. Double-click it
3. Add your tasks
4. Let it manage your tasks for you :relaxed:

And it's FREE! :trollface:

Features:

- [x] Managing and marking tasks
- [ ] Undo most recent change (coming soon!)

If you're a Java programmer, you can use it to practice Java too. Here's the main method:

```ruby
public class Launcher {
    public static void main(String[] args) {
        Application.launch(EchoBot.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

JasonRay168 committed:
Rename, Greet, Exit

JasonRay168 committed:
Rename, Greet, Exit

JasonRay168 committed:
Echo

JasonRay168 committed:
Add, List

JasonRay168 committed:
Mark and Unmark as Done

JasonRay168 committed:
ToDos, Events, Deadlines

JasonRay168 committed:
Automated Text UI Testing

JasonRay168 committed:
Level 5. Handle Errors

JasonRay168 committed:
Delete

JasonRay168 committed:
Change tabs into spaces and rename the chatbot to Duke

JasonRay168 committed:
Store task list in hard drive

JasonRay168 committed:
Merge branch-Level-7

JasonRay168 committed:
Support date inputs for deadline and event tasks

JasonRay168 committed:
Merge branch-Level-8

JasonRay168 committed:
Separate Duke class main functions to several classes

JasonRay168 committed:
Create packages for classes

JasonRay168 committed:
Move most error handling from Ui to other classes

JasonRay168 committed:
Merge branch 'add-gradle-support'

JasonRay168 committed:
Merged add-gradle-support branch

JasonRay168 committed:
Add gradle support to execute run and build

JasonRay168 committed:
Add JUnit tests for Parser and Task

JasonRay168 committed:
Merge branch-A-Gradle

JasonRay168 committed:
Merge master branch

JasonRay168 committed:
Merge branch-A-JUnit

JasonRay168 committed:
Fix faulty runtest.bat

JasonRay168 committed:
Fix runtest.bat compilation errors and update EXPECTED.TXT

JasonRay168 committed:
Add JavaDoc comments to java files under duke package

JasonRay168 committed:
Make code follow coding standard

JasonRay168 committed:
Allow users to search for task names by searching for a keyword

JasonRay168 committed:
Merge branch-A-JavaDoc

JasonRay168 committed:
Handle merge conflicts after merging A-JavaDoc and A-CodingStandard branches

JasonRay168 committed:
Handle merge conflicts after merging branch-Level-9

JasonRay168 committed:
Declare multiple helper methods to handle user input in Ui class

JasonRay168 committed:
Add JavaFX, Launcher, and update Gradle build file

JasonRay168 committed:
Allow Duke to echo the user input

JasonRay168 committed:
Fix current checkstyle violations

JasonRay168 committed:
Merge branch-A-CheckStyle

JasonRay168 committed:
Merge master branch

JasonRay168 committed:
Add dialog boxes and stock pictures

JasonRay168 committed:
Flip Duke's chat outputs

JasonRay168 committed:
Changed package name to echobot and class name to EchoBot

JasonRay168 committed:
Add FXML support

JasonRay168 committed:
Update FXML tag

JasonRay168 committed:
Fix issues with FXML support

JasonRay168 committed:
Edit MainWindow and DialogBox FXML configurations

JasonRay168 committed:
Merge branch-Level-10

czhiruo commented on src/main/java/echobot/utilities/Type.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Enum class to declare types of tasks
+ */
+public enum Type {
```
I like how you have ENUMS to set predefined types

czhiruo commented on src/main/java/echobot/utilities/Ui.java:
```diff
@@ -0,0 +1,245 @@
+/**
+ * Handles user interaction
+ */
+public class Ui {
```
I like how the UI class displays all the strings to be printed to the user.

czhiruo commented on src/main/java/echobot/utilities/Parser.java:
```diff
@@ -0,0 +1,92 @@
+     * @param ui Ui where the helper functions are located
+     * @return True or false signifying breaking or continuing the loop
+     */
+    public String handleInput(TaskList tasks, Input input, Ui ui) {
```
I like how this method is clear and concise in abstracting the input and output

czhiruo commented on src/main/java/echobot/utilities/Task.java:
```diff
@@ -0,0 +1,78 @@
+     *
+     * @return the String representation of the task
+     */
+    public String convertToString() {
```
Any reason why you did it this way instead of overriding the toString method?

JasonRay168 replied:
Simply put, I only remembered I can overwrite toString() after about a week implementing this, and at that point I didn't want to change the method fearing that the literal string representation of the tasks might be used in the future.

Turns out It was needed to extract it into the file in the hard drive, but at this point, I don't see any difference in coding implementation. It's just that the file will become more readable that way.

Which is why I had raised a concern about the file being unreadable in the forums hahaha

I'll be sure to remember this for the tP, thanks

czhiruo reviewed (commented):
Overall, your code is very well abstracted and clear

JasonRay168 committed:
Change error message sligthly to prepare for JAR release

JasonRay168 committed:
Add assertions to check if text file exists

There might be instances where the txt file might be accidentally delete by user.

Hence, assertions is added to assert that the file exists when loading the file and when overwriting the file.

JasonRay168 committed:
Improve code quality

Some fields and methods are unused, some methods are just too long, and some declarations are obsolete.

Thus, removing unused and obsolete parts and shortening down methods to adhere to SLAP will improve code quality.

JasonRay168 committed:
Merge branch-A-Assertions

JasonRay168 committed:
Merge master branch

JasonRay168 committed:
Merge branch-A-CodeQuality

JasonRay168 committed:
Merge master branch

JasonRay168 committed:
Merge branch 'master' of https://github.com/JasonRay168/ip into branch-A-CodeQuality

JasonRay168 committed:
Handle small mistakes in code

JasonRay168 committed:
Merge pull request #4 from JasonRay168/branch-A-CodeQuality

Merge branch-A-CodeQuality

JasonRay168 committed:
Add an undo feature to undo most recent change

There might be some cases where users will accidentally delete or change the content of the list.

In such cases, an undo feature will be really helpful in helping users retrieve lost or changed information.

The undo feature can only allow users to undo the most recent change and cannot be used to undo more than one change.

Furthermore, the undo feature will only detect mark, unmark, delete, and the three add task commands as commands that change the content of the list. After all, the other features such as list and find does not change the contents of the task list.

Finally, upon exiting and restarting the chatbot, the undo will not work. It will only work when a change is made in the current session.

JasonRay168 committed:
Merge branch-BCD-Extension

JasonRay168 committed:
Refactor JavaDoc comments to follow coding standard

JasonRay168 committed:
Add a user guide and finalize features

JasonRay168 committed:
Merge branch-A-UserGuide

JasonRay168 committed:
Edit error message when tasks.txt file is not found

JasonRay168 committed:
Include product name on top of user guide

JasonRay168 committed:
Merge branch-A-UserGuide after small change

JasonRay168 committed:
Add name of the app on top of main window
